File: src/utils/SerializationHeader.cpp
//
// Created by Andrei on 15-May-21.
//

#include <comm/utils/SerializationHeader.h>
#include <comm/utils/NetworkIncludes.h>

using namespace comm;
// using namespace std;

SerializationHeader::SerializationHeader(bool create) : localBuffer(nullptr), passedBuffer(nullptr), created(create) {
    if (this->created) {
        this->localBuffer = new char[4];
    }
}

SerializationHeader::SerializationHeader(char *buffer) : SerializationHeader(false) {
    this->localBuffer = buffer;
}

SerializationHeader::SerializationHeader(Buffer &buffer) : SerializationHeader(false) {
    this->passedBuffer = &buffer;
}

SerializationHeader::SerializationHeader(int header) : SerializationHeader() {
    this->setInt(header);
}

SerializationHeader::SerializationHeader(unsigned char serializationIteration, unsigned char sendIteration,
                                         unsigned short sendSize) : SerializationHeader() {
    this->setData(serializationIteration, sendIteration, sendSize);
}

SerializationHeader::~SerializationHeader() {
    if (this->created) {
        delete[] this->localBuffer;
    }
}

void SerializationHeader::reset() {
    this->setInt(0);
}

void SerializationHeader::setInt(int header, bool local) {
    this->setSendSize((unsigned short) (header & 65535), local);  // mod 2^16
    header >>= 16;  // div 2^16
    this->setSendIteration((unsigned char) (header & 255));  // mod 2^8
    this->setSerializationIteration((unsigned char) (header >> 8));  // div 2^8
}

void SerializationHeader::setData(unsigned char _serializationIteration, unsigned char _sendIteration,
                                  unsigned short _sendSize, bool local) {
    this->setSerializationIteration(_serializationIteration);
    this->setSendIteration(_sendIteration);
    this->setSendSize(_sendSize, local);
}

void SerializationHeader::setSerializationIteration(unsigned char _serializationIteration) {
    if (this->passedBuffer != nullptr) {
        this->localBuffer[0] = (char) _serializationIteration;
    } else {
        this->passedBuffer->setChar((char) _serializationIteration, 0);
    }
}

void SerializationHeader::setSendIteration(unsigned char _sendIteration) {
    if (this->passedBuffer != nullptr) {
        this->localBuffer[1] = (char) _sendIteration;
    } else {
        this->passedBuffer->setChar((char) _sendIteration, 1);
    }
}

void SerializationHeader::setSendSize(unsigned short _sendSize, bool setLocal) {
    auto sendSize = (short) ((setLocal) ? _sendSize : ntohs(_sendSize));
    if (this->passedBuffer != nullptr) {
        memcpy(this->localBuffer + 2, &sendSize, 2);
    } else {
        this->passedBuffer->setShort(sendSize, 2);
    }
}

void SerializationHeader::setBuffer(char *buffer, bool setLocal, int start) const {
    buffer[start] = (char) this->getSerializationIteration();
    buffer[start + 1] = (char) this->getSendIteration();
    auto x = this->getSendSize(setLocal);
    memcpy(buffer + start + 2, &x, 2);
}

void SerializationHeader::setBuffer(Buffer &buffer, bool setLocal, int start) const {
    buffer.setInt(this->getInt(setLocal), start);
}

unsigned char SerializationHeader::getSerializationIteration() const {
    if (this->passedBuffer != nullptr) {
        return this->localBuffer[0];
    } else {
        return this->passedBuffer->getChar(0);
    }
}

unsigned char SerializationHeader::getSendIteration() const {
    if (this->passedBuffer != nullptr) {
        return this->localBuffer[1];
    } else {
        return this->passedBuffer->getChar(1);
    }
}

unsigned short SerializationHeader::getSendSize(bool getLocal) const {
    unsigned short sendSize;
    if (this->passedBuffer != nullptr) {
        memcpy(&sendSize, this->localBuffer + 2, 2);
    } else {
        sendSize = this->passedBuffer->getShort(2);
    }
    if (getLocal) {
        return sendSize;
    } else {
        return htons(sendSize);
    }
}

int SerializationHeader::getInt(bool getLocal) const {
    return (this->getSerializationIteration() << 24) + (this->getSendIteration() << 16) + this->getSendSize(getLocal);
}

File: csharp/Comm/utils/SerializationHeader.cs
using System;
using System.Net;

namespace Comm.utils {
    public class SerializationHeader {
        public SerializationHeader(bool create = true) {
            this.created = create;
            if (this.created) {
                this.localBuffer = new byte[4];
            }
        }

        public SerializationHeader(ref byte[] buffer) {
            this.localBuffer = buffer;
        }

        public SerializationHeader(Buffer buffer) {
            this.passedBuffer = buffer;
        }

        public SerializationHeader(int header) {
            this.setInt(header);
        }

        public SerializationHeader(byte serializationIteration, byte sendIteration, ushort sendSize) {
            this.setData(serializationIteration, sendIteration, sendSize);
        }

        ~SerializationHeader() {
            if (this.created) {
                this.localBuffer = null;
            }
        }

        public void reset() {
            this.setInt(0);
        }

        public void setInt(int header, bool local = true) {
            this.setSendSize((ushort)(header & 65535), local);  // mod 2^16
            header >>= 16;  // div 2^16
            this.setSendIteration((byte)(header & 255));  // mod 2^8
            this.setSerializationIteration((byte)(header >> 8));  // div 2^8
        }

        public void setData(byte _serializationIteration, byte _sendIteration, ushort _sendSize, bool local = true) {
            this.setSerializationIteration(_serializationIteration);
            this.setSendIteration(_sendIteration);
            this.setSendSize(_sendSize, local);
        }

        public void setSerializationIteration(byte _serializationIteration) {
            if (this.passedBuffer != null) {
                this.localBuffer[0] = _serializationIteration;
            } else {
                this.passedBuffer.setByte(_serializationIteration, 0);
            }
        }

        public void setSendIteration(byte _sendIteration) {
            if (this.passedBuffer != null) {
                this.localBuffer[1] = _sendIteration;
            } else {
                this.passedBuffer.setByte(_sendIteration, 1);
            }
        }

        public void setSendSize(ushort _sendSize, bool setLocal = true) {
            short sendSize = (short)((setLocal) ? _sendSize : IPAddress.NetworkToHostOrder(_sendSize));
            if (this.passedBuffer != null) {
                utils.Utils.memcpy(this.localBuffer, 2, BitConverter.GetBytes(sendSize), 0, 2);
            } else {
                this.passedBuffer.setShort(sendSize, 2);
            }
        }

        public void setBuffer(byte[] buffer, bool setLocal, int start = 0) {
            buffer[start] = this.getSerializationIteration();
            buffer[start + 1] = this.getSendIteration();
            ushort x = this.getSendSize(setLocal);
            utils.Utils.memcpy(buffer, (ulong)start + 2, BitConverter.GetBytes(x), 0, NetworkData.shortSize);
        }

        public void setBuffer(utils.Buffer buffer, bool setLocal, int start = 0) {
            buffer.setInt(this.getInt(setLocal), start);
        }

        public byte getSerializationIteration() {
            if (this.passedBuffer != null) {
                return this.localBuffer[0];
            } else {
                return this.passedBuffer.getByte(0);
            }
        }

        public byte getSendIteration() {
            if (this.passedBuffer != null) {
                return this.localBuffer[1];
            } else {
                return this.passedBuffer.getByte(1);
            }
        }

        public ushort getSendSize(bool getLocal = true) {
            ushort sendSize;
            if (this.passedBuffer != null) {
                sendSize = (ushort)BitConverter.ToInt16(this.localBuffer, 2);
            } else {
                sendSize = (ushort)this.passedBuffer.getShort(2);
            }
            if (getLocal) {
                return sendSize;
            } else {
                return (ushort)IPAddress.HostToNetworkOrder((short)sendSize);
            }
        }

        public int getInt(bool getLocal = true) {
            return (this.getSerializationIteration() << 24) + (this.getSendIteration() << 16) + this.getSendSize(getLocal);
        }

        private byte[] localBuffer;
        private bool created;
        private utils.Buffer passedBuffer;
    }
}
